example/bwmf/bwmf_task.go

```diff
@@ -31,168 +33,516 @@ Topology: the topology is different from task to task. Each task will consider i
+type event struct {
+	ctx      context.Context
+	epoch    uint64
+	fromID   uint64
+	request  *pb.Request
+	retT     chan *pb.DenseMatrixShard
+	retD     chan *pb.DenseMatrixShard
+	response proto.Message
+	// to be extended
+}
 
-	go func() {
-		if t.epoch%2 == 0 {
-			b, err = json.Marshal(t.dShard)
+func (bt *bwmfTask) run() {
+	for {
+		select {
+		case ec := <-bt.epochChange:
+			bt.logger.Println("SELECT- Epoch change to ", ec.epoch, " at task ", bt.taskID)
+			bt.doEnterEpoch(ec.ctx, ec.epoch)
+
+		case reqT := <-bt.getT:
+			bt.logger.Println("SELECT- GetT at epoch", bt.epoch)
+			// XXX: tShard is guaranteed to have been updated here.
+			err := bt.framework.CheckGRPCContext(reqT.ctx)
 			if err != nil {
-				t.logger.Fatalf("Slave can't encode dShard error: %v\n", err)
+				bt.logger.Panicf("GetTShard grpc broken in context at task %d for epoch %d.", bt.taskID, bt.epoch)
+				close(reqT.retT)
 			}
-		} else {
-			b, err = json.Marshal(t.tShard)
+			bt.logger.Printf("To respond with retT at task %d epoch %d.", bt.taskID, bt.epoch)
+			reqT.retT <- bt.shardedT
+			bt.logger.Printf("Responded with retT at task %d epoch %d.", bt.taskID, bt.epoch)
+		case reqD := <-bt.getD:
+			bt.logger.Println("SELECT- GetD at epoch", bt.epoch)
+			// XXX: tShard is guaranteed to have been updated here.
+
+			bt.logger.Println("Before checking GRPC status...")
+			bt.logWriter.Flush()
+			err := bt.framework.CheckGRPCContext(reqD.ctx)
+			bt.logger.Println("After checking GRPC status: ", err)
+			bt.logWriter.Flush()
 			if err != nil {
-				t.logger.Fatalf("Slave can't encode tShard error: %v\n", err)
+				bt.logger.Panicf("GetDShard grpc broken in context at task %d for epoch %d.", bt.taskID, bt.epoch)
+				close(reqD.retD)
+				continue
 			}
+			bt.logger.Printf("To respond with retD at task %d epoch %d.", bt.taskID, bt.epoch)
+			reqD.retD <- bt.shardedD
+			bt.logger.Printf("Responded with retD at task %d epoch %d.", bt.taskID, bt.epoch)
+		case done := <-bt.updateDone:
+			bt.logger.Println("SELECT- update t/d is done at epoch", bt.epoch)
+			go bt.checkpoint(bt.epoch)
+			// bt.framework.FlagMeta(done.ctx, "toMaster", "metaReady")
+			bt.logger.Println("Flagging metaReady.")
+			bt.framework.FlagMeta(done.ctx, "Neighbors", "metaReady")
+		case dr := <-bt.dataReady:
+			bt.logger.Println("SELECT- DataReady at epoch", bt.epoch)
+			// each event comes as the shard is ready
+			resp, bOk := dr.response.(*pb.Response)
+			if !bOk {
+				bt.logger.Fatalf("Cannot convert proto message to Response: %v", dr.response)
+			}
+			// In even epoch, we fetch all D shards and update T; while in odd epoch, we fetch T and update D.
+			if bt.epoch%2 == 0 {
+				_, exists := bt.fullD[resp.BlockId]
+				if exists {
+					bt.logger.Fatalf("Duplicated response from block %d for full D shard  with response: %v", resp.BlockId, dr.response)
+				}
+				bt.fullD[resp.BlockId] = resp.Shard
+				// if all tShards have arrived
+				if len(bt.fullD) == int(bt.numOfTasks) {
```

Comment:
FullTopo now includes all tasks as neighbors incliding the task itself. So the DShard on tjis task will also come via DataReady.